client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java

```diff
@@ -503,36 +567,31 @@ protected void limitZeroInFlight(String mapKey, PushState pushState) throws IOEx
     }
   }
 
-  private boolean waitRevivedLocation(
-      ConcurrentHashMap<Integer, PartitionLocation> map, int partitionId, int epoch) {
+  boolean checkRevivedLocation(int shuffleId, int mapId, int partitionId, int epoch, boolean wait) {
+    ConcurrentHashMap<Integer, PartitionLocation> map = reducePartitionMap.get(shuffleId);
     PartitionLocation currentLocation = map.get(partitionId);
-    if (currentLocation != null && currentLocation.getEpoch() > epoch) {
+    if (mapperEnded(shuffleId, mapId)
+        || (currentLocation != null && currentLocation.getEpoch() > epoch)) {
       return true;
-    }
-
-    long sleepTimeMs = RND.nextInt(50);
-    if (sleepTimeMs > 30) {
-      try {
-        TimeUnit.MILLISECONDS.sleep(sleepTimeMs);
-      } catch (InterruptedException e) {
-        logger.error("Waiting revived location was interrupted.", e);
-        Thread.currentThread().interrupt();
+    } else if (wait) {
+      long sleepTimeMs = RND.nextInt(50);
+      if (sleepTimeMs > 30) {
+        try {
+          TimeUnit.MILLISECONDS.sleep(sleepTimeMs);
+        } catch (InterruptedException e) {
+          logger.error("Waiting revived location was interrupted.", e);
+          Thread.currentThread().interrupt();
```

Comment:
Interrupt here should change refer to https://github.com/apache/incubator-celeborn/pull/1486